### PR TITLE
[WIP] Handle self-loops(needs to be tested)

### PR DESCRIPTION
**Pull request by idetatsu**
I am still running the regression test to make sure there is no unwanted changes, but here's the code that we discussed last week.

It allows both graphs g1 and g2 to have self-loops and (hopefully) is able to align them as well. We'll be able to test if that's the case by the test suite that I am working on right now.

### CODE REVIEW

File: src/Graph.cpp [hunk 1]
```diff
@@ -1166,6 +1152,14 @@ unordered_map<uint,string> Graph::getIndexToNodeNameMap() const {
     return res;
 }
 
+bool Graph::hasSelfLoop(uint source) const {
+    vector<uint> neighbors = adjLists[source];
+    if(std::find(neighbors.begin(), neighbors.end(), source) != neighbors.end()) {
```
Comment (waynebhayes) [inline src/Graph.cpp:1195]: This seems terribly expensive. Why not just check the single bit that's along the diagonal of the adjacency matrix?

Note that the method you have here *will* be necessary if SPARSE is true; but if SPARSE is false then just perform the O(1) operation of looking at the diagonal of the adjacency matrix.

File: src/methods/SANA.cpp [hunk 2]
```diff
@@ -1428,10 +1428,14 @@ bool SANA::scoreComparison(double newAligEdges, double newInducedEdges, double n
 
 int SANA::aligEdgesIncChangeOp(uint source, uint oldTarget, uint newTarget) {
     int res = 0;
-    const uint n = G1AdjLists[source].size();
-    uint neighbor;
-    for (uint i = 0; i < n; ++i) {
-        neighbor = G1AdjLists[source][i];
+
+    vector<uint> v = G1AdjLists[source];
+    if(G1->hasSelfLoop(source)) {
```
Comment (waynebhayes) [inline src/methods/SANA.cpp]: Again even calling this function may add uncessary expense. If SPARSE is false, just check the adjacencry matrix inline right here. Perform the function call only of SPARSE is true and it's necessary to perform the O(degree) search.

File: src/methods/SANA.cpp [hunk 3]
```diff
@@ -1440,25 +1444,32 @@ int SANA::aligEdgesIncChangeOp(uint source, uint oldTarget, uint newTarget) {
...
 int SANA::aligEdgesIncSwapOp(uint source1, uint source2, uint target1, uint target2) {
     int res = 0;
-    const uint n = G1AdjLists[source1].size();
-    uint neighbor;
-    uint i = 0;
-    for (; i < n; ++i) {
-        neighbor = G1AdjLists[source1][i];
+
+    vector<uint> v1 = G1AdjLists[source1];
+    if(G1->hasSelfLoop(source1)) {
```
Comment (waynebhayes) [inline src/methods/SANA.cpp]: Same here.

File: src/methods/SANA.cpp [hunk 4]
```diff
@@ -1440,25 +1444,32 @@ int SANA::aligEdgesIncChangeOp(uint source, uint oldTarget, uint newTarget) {
...
+    for(uint neighbor : v1) {
         res -= G2Matrix[target1][(*A)[neighbor]];
         res += G2Matrix[target2][(*A)[neighbor]];
     }
-    const uint m = G1AdjLists[source2].size();
-    for (i = 0; i < m; ++i) {
-        neighbor = G1AdjLists[source2][i];
+
+    vector<uint> v2 = G1AdjLists[source2];
+    if(G1->hasSelfLoop(source2)) {
```
Comment (waynebhayes) [inline src/methods/SANA.cpp]: And here.

**Review by waynebhayes — CHANGES_REQUESTED** [review]
I think your code is *correct* but may be unnecessarily expensive. See comments for details.

Especially in the IncChangeOp() dfunctions, these get called millions of times per second. They need to be lightning fast.

File: src/Graph.cpp (hunk 1 above)
Comment (idetatsu) [reply src/Graph.cpp:1195]: I had just forgotten I could use adjacency matrix when I implemented it. I'll fix this.

**Comment by waynebhayes** [comment]
Ok cool

**Comment by waynebhayes** [comment]
Any more news on this?

**Comment by waynebhayes** [comment]
It's been almost a month.  Can we get this in soon?  I also haven't seen
you in meetings for awhile.  If you want to continue in the group, get this
done this week.

--
Wayne Hayes, Ph.D.
Associate Professor of Computer Science, University of California, Irvine
Director, UCI-SDSU Joint Ph.D. Program in Computational Science (UCI side)
*The evidence is now incontrovertible that Russia interfered with the 2016
US Presidential election by manipulating public opinion with disinformation
via thousands of fake social media accounts. Why? Follow the money: Trump
has been laundering it for Russian billionaires, including Putin, for 20
years. Details and fact checking
at https://secure.avaaz.org/campaign/en/shocking_truth_about_trump/?bebhylb
<https://secure.avaaz.org/campaign/en/shocking_truth_about_trump/?bebhylb>.*
The opinions expressed in this email by Professor Hayes are his and his
alone, and are not meant to represent those of the University of California.

**Comment by idetatsu** [comment]
@waynebhayes I'm sorry I took so long to finish this. I have made changes regarding what you mentioned earlier. Please check it out.

**Review by waynebhayes — APPROVED** [review]

**Pull request merged**